src/collisions.rs

```diff
@@ -0,0 +1,462 @@
+        }
+
+        bounds_point
+    }
+
+    /// Returns (Closest, EdgeNormal)
+    pub fn closest_point_on_border(&self, point: Vec2) -> (Vec2, Vec2) {
+        let mut edge_normal = Vec2::ZERO;
+
+        let mut res = Vec2::new(0.0, 0.0);
+        res.x = point.x.clamp(self.left(), self.right());
+        res.y = point.y.clamp(self.top(), self.bottom());
+
+        if self.contains(res) {
+            let dl = res.x - self.left();
+            let dr = self.right() - res.x;
+            let dt = res.y - self.top();
+            let db = self.bottom() - res.y;
+
+            let min = dl.min(dr.min(dt.min(db)));
+            if min == dt {
+                res.y = self.top();
+                edge_normal.y = -1.0;
+            } else if min == db {
+                res.y = self.bottom();
+                edge_normal.y = 1.0;
+            } else if min == dl {
+                res.x = self.left();
+                edge_normal.x = -1.0;
+            } else {
+                res.x = self.right();
+                edge_normal.x = 1.0;
+            }
+        } else {
+            if res.x == self.left() {
+                edge_normal.x = -1.0;
+            }
+            if res.x == self.right() {
+                edge_normal.x = 1.0;
+            }
+            if res.y == self.top() {
+                edge_normal.y = -1.0;
+            }
+            if res.y == self.bottom() {
+                edge_normal.y = 1.0;
+            }
+        }
+
+        (res, edge_normal)
+    }
+
+    fn ray_intersects(&self, ray: &Ray2D) -> Option<f32> {
+        let mut distance = 0.0;
+        let mut max = f32::MAX;
+
+        if ray.direction.x.abs() < 1e-06 {
+            if ray.start.x < self.x || ray.start.x > self.x + self.width {
+                return None;
+            }
+        } else {
+            let inv_x = 1.0 / ray.direction.x;
+            let mut left = (self.x - ray.start.x) * inv_x;
+            let mut right = (self.x + self.width - ray.start.x) * inv_x;
+            if left > right {
+                let tmp = left;
+                left = right;
+                right = tmp;
+            }
+
+            distance = left.max(distance);
+            max = right.min(max);
+            if distance > max {
+                return None;
+            }
+        }
+
+        if ray.direction.y.abs() < 1e-06 {
+            if ray.start.y < self.y || ray.start.y > self.y + self.height {
+                return None;
+            }
+        } else {
+            let inv_y = 1.0 / ray.direction.y;
+            let mut top = (self.y - ray.start.y) * inv_y;
+            let mut bottom = (self.y + self.height - ray.start.y) * inv_y;
+            if top > bottom {
+                let tmp = top;
+                top = bottom;
+                bottom = tmp;
+            }
+
+            distance = top.max(distance);
+            max = bottom.min(max);
+            if distance > max {
+                return None;
+            }
+        }
```

Comment:
we probably should move it in function probably because it is the same 20 lines of code as for x direction

Reply:
Fixed